test/system/src/test/java/io/pravega/test/system/BasicCBRTest.java

```diff
@@ -0,0 +1,203 @@
+        Service conService = Utils.createPravegaControllerService(null);
+        List<URI> ctlURIs = conService.getServiceDetails();
+        controllerURI = ctlURIs.get(0);
+
+        final ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+
+        controller = new ControllerImpl(ControllerImplConfig.builder()
+                .clientConfig(clientConfig)
+                .maxBackoffMillis(5000).build(), executor);
+        streamManager = StreamManager.create(clientConfig);
+
+        assertTrue("Creating scope", streamManager.createScope(SCOPE));
+        assertTrue("Creating stream", streamManager.createStream(SCOPE, STREAM,
+                StreamConfiguration.builder()
+                        .scalingPolicy(ScalingPolicy.fixed(1))
+                        .retentionPolicy(RetentionPolicy.bySizeBytes(MIN_SIZE_IN_STREAM, MAX_SIZE_IN_STREAM)).build()));
+    }
+
+    @After
+    public void tearDown() {
+        streamManager.close();
+        ExecutorServiceHelpers.shutdown(executor);
+    }
+
+    @Test
+    public void basicCBRTest() throws Exception {
+        final ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+
+        @Cleanup
+        EventStreamClientFactory clientFactory = EventStreamClientFactory.withScope(SCOPE, clientConfig);
+        @Cleanup
+        EventStreamWriter<String> writer = clientFactory.createEventWriter(STREAM, new JavaSerializer<>(),
+                EventWriterConfig.builder().build());
+
+        // Write a single event.
+        log.info("Writing event e1 to {}/{}", SCOPE, STREAM);
+        writer.writeEvent("e1", SIZE_30_EVENT).join();
+
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(SCOPE, clientConfig);
+        readerGroupManager.createReaderGroup(READER_GROUP, ReaderGroupConfig.builder()
+                .retentionType(ReaderGroupConfig.StreamDataRetention.MANUAL_RELEASE_AT_USER_STREAMCUT)
+                .disableAutomaticCheckpoints()
+                .stream(Stream.of(SCOPE, STREAM)).build());
+        ReaderGroup readerGroup = readerGroupManager.getReaderGroup(READER_GROUP);
+        @Cleanup
+        EventStreamReader<String> reader = clientFactory.createReader(READER_GROUP + "-" + 1,
+                READER_GROUP, new JavaSerializer<>(), readerConfig);
+
+        // Read one event.
+        log.info("Reading event e1 from {}/{}", SCOPE, STREAM);
+        EventRead<String> read = reader.readNextEvent(READ_TIMEOUT);
+        assertFalse(read.isCheckpoint());
+        assertEquals("data of size 30", read.getEvent());
+
+        // Update the retention stream-cut.
+        log.info("{} generating stream-cuts for {}/{}", READER_GROUP, SCOPE, STREAM);
+        CompletableFuture<Map<Stream, StreamCut>> futureCuts = readerGroup.generateStreamCuts(streamCutExecutor);
+        // Wait for 5 seconds to force reader group state update.
+        Exceptions.handleInterrupted(() -> TimeUnit.SECONDS.sleep(5));
```

Comment:
I don't like the idea of adding sleeps to tests as it increases possibility of making tests flaky. Why do we need to wait for reader group state update here?

Reply:
Calling `generateStreamCuts` leads to the creation of a silent checkpoint. In order for the ReaderGroup to process the checkpoint we need to wait and allow the ReaderGroupState to update and fetch the checkpoint so that it can be processed. This is present in the form of a timer (`checkpointTimer`) in the `ReaderGroupStateManager`.